bgpd/bgp_nexthop.c

```diff
@@ -453,21 +486,88 @@ static void bgp_connected_cleanup(struct route_table *table,
 	}
 }
 
-int bgp_nexthop_self(struct bgp *bgp, struct in_addr nh_addr)
+int bgp_nexthop_self(struct bgp *bgp, afi_t afi, uint8_t type, uint8_t sub_type,
+		struct attr *attr, struct bgp_node *rn)
 {
-	struct bgp_addr tmp, *addr;
-	struct tip_addr tmp_tip, *tip;
-
-	tmp.addr = nh_addr;
+	struct prefix p;
+	afi_t new_afi = afi;
+	struct bgp_addr tmp_addr, *addr = NULL;
+	struct tip_addr tmp_tip, *tip = NULL;
+
+	int is_bgp_static_route = ((type == ZEBRA_ROUTE_BGP)
+			&& (sub_type == BGP_ROUTE_STATIC))
+			? 1
+			: 0;
+
+	memset(&p, 0, sizeof(struct prefix));
+	memset(&tmp_addr, 0, sizeof(struct bgp_addr));
+
+	if (!is_bgp_static_route)
+		new_afi = BGP_ATTR_NEXTHOP_AFI_IP6(attr) ? AFI_IP6 : AFI_IP;
+
+	switch (new_afi) {
+	case AFI_IP:
+		p.family = AF_INET;
+		if (is_bgp_static_route) {
+			p.u.prefix4 = rn->p.u.prefix4;
+			p.prefixlen = rn->p.prefixlen;
+		} else {
+			/* Here we need to find out which nexthop to be used*/
+			if (attr->flag &
+					ATTR_FLAG_BIT(BGP_ATTR_NEXT_HOP)) {
+
+				p.u.prefix4 = attr->nexthop;
+				p.prefixlen = IPV4_MAX_BITLEN;
+
+			} else if ((attr->mp_nexthop_len) &&
+					((attr->mp_nexthop_len ==
+					  BGP_ATTR_NHLEN_IPV4) ||
+					 (attr->mp_nexthop_len ==
+					  BGP_ATTR_NHLEN_VPNV4))) {
+				p.u.prefix4 =
+					attr->mp_nexthop_global_in;
+				p.prefixlen = IPV4_MAX_BITLEN;
+			} else
+				return 0;
+		}
+		break;
+	case AFI_IP6:
+		p.family = AF_INET6;
+
+		if (is_bgp_static_route) {
+			p.u.prefix6 = rn->p.u.prefix6;
+			p.prefixlen = rn->p.prefixlen;
+		} else {
+			p.u.prefix6 = attr->mp_nexthop_global;
```

Comment:
what if next-hop is a link-local address?

Reply:
as per code above in bgp_connected_add(), link local addresses are not taken into account.
dont know if it is good or not.
I mean, we should always use bgp ipv6 working with unnumbered interfaces. ( see https://github.com/FRRouting/frr/pull/4199 )

Reply:
Actually link-local address be it ipv4 or ipv6 was never part of this connected hash table. So i have not added it in this pull request dealing with mainly optimisation.
We had an internal discussion at vmware on the same topic. Since this table is used for various validation, need to do various test to confirm it would not create other issue. So we decided to tract this as separate issue outside this pull request. Pls let me know your thoughts too.

Reply:
ok for me.thanks